src/java.base/share/classes/java/lang/ClassLoader.java

```diff
@@ -1049,14 +1050,22 @@ protected final Class<?> defineClass(String name, java.nio.ByteBuffer b,
 
         protectionDomain = preDefineClass(name, protectionDomain);
         String source = defineClassSourceLocation(protectionDomain);
-        Class<?> c = defineClass2(this, name, b, b.position(), len, protectionDomain, source);
-        postDefineClass(c, protectionDomain);
-        return c;
+
+        SharedSecrets.getJavaNioAccess().acquireSession(b);
+        try {
+            Class<?> c = defineClass2(this, name, b, b.position(), len, protectionDomain, source);
+            postDefineClass(c, protectionDomain);
```

Comment:
Should we leave postDefineClass out of this acquire-release scope? I don't see any reason including this.

Reply:
I don't think it matters here because something looking to close the arena around the time that it wants to defineClass with memory allocated from that arena is broken.